shared-libs/rules-engine/src/contact-state-store.js

```diff
@@ -0,0 +1,209 @@
+
+      onStateChange(state);
+      return true;
+    }
+
+    return false;
+  },
+
+  /**
+   * @param {int} calculatedAt Timestamp of the calculation
+   * @param {string[]} contactIds Array of contact ids to be marked as freshly calculated
+   */
+  markFresh: (calculatedAt, contactIds) => {
+    if (!Array.isArray(contactIds)) {
+      contactIds = [contactIds];
+    }
+    contactIds = contactIds.filter(id => id);
+
+    if (contactIds.length === 0) {
+      return Promise.resolve();
+    }
+
+    for (let contactId of contactIds) {
+      state.contactStates[contactId] = { calculatedAt };
+    }
+
+    return onStateChange(state);
+  },
+
+  /**
+   * @param {string[]} contactIds Array of contact ids to be marked as dirty
+   */
+  markDirty: contactIds => {
+    if (!Array.isArray(contactIds)) {
+      contactIds = [contactIds];
+    }
+    contactIds = contactIds.filter(id => id);
+
+    if (contactIds.length === 0) {
+      return Promise.resolve();
+    }
+
+    for (let contactId of contactIds) {
+      if (!state.contactStates[contactId]) {
+        state.contactStates[contactId] = {};
+      }
+
+      state.contactStates[contactId].isDirty = true;
+    }
+
+    return onStateChange(state);
+  },
+
+  /**
+   * @returns {string[]} The id of all contacts tracked by the store
+   */
+  getContactIds: () => Object.keys(state.contactStates),
+
+  /**
+   * The rules system supports the concept of "headless" reports and "headless" task documents. In these scenarios, a report exists on a user's device while the associated
+   * contact document of that report is not on the device. A common scenario associated with this case is during supervisor workflows where supervisors sync reports with the 
+   * needs_signoff attribute but not the associated patient.
+   * 
+   * In these cases, getting a list of "all the contacts with rules" requires us to look not just through contact docs, but also through reports. To avoid this costly operation, 
+   * the contact-state-store maintains a flag which indicates if the contact ids in the store can serve as a trustworthy authority.
+   * 
+   * markAllFresh should be called when the list of contact ids within the store is the complete set of contacts with rules
+   */
+  markAllFresh: (calculatedAt, contactIds) => {
+    state.allContactIds = true;
+    return self.markFresh(calculatedAt, contactIds);
+  },
+
+  /**
+   * @returns True if markAllFresh has been called on the current store state.
+   */
+  hasAllContacts: () => !!state.allContactIds,
+
+  /**
+   * @returns {string} User contact document
+   */
+  currentUser: () => currentUser,
+};
+
+const hashRulesConfig = (settingsDoc, userDoc, salt) => {
+  const rulesConfig = {
+    version: '1', // change this release-to-release to force task recalculation after deployment
+    rules: settingsDoc && settingsDoc.tasks && settingsDoc.tasks.rules,
+    targets: settingsDoc && settingsDoc.targets,
+    userDoc,
+    salt,
```

Comment:
This is for the "recalculate rules" button that I was hoping to add beside the "debug mode" button on the about page. I think it would be quite useful for testing.

Reply:
I think it would be clearer to have a `force` parameter which, if set, doesn't run the diff at all.

Reply:
Just removed for now